File: frontend/src/constants/disciplines.js
/**
 * Unified Discipline Constants
 * Used across the application for consistency in Actions, Tasks, AI Recommendations, and FMEA.
 * 
 * CONFIGURATION: To change disciplines, update:
 * 1. This file (frontend constants)
 * 2. Backend: /app/backend/iso14224_models.py (Discipline enum)
 * 3. EquipmentTypeItem.jsx DISCIPLINE_COLORS for visual styling
 */

// Standard disciplines - aligned across the application
export const DISCIPLINES = [
  { value: "rotating", label: "Rotating", color: "bg-blue-100 text-blue-700" },
  { value: "static", label: "Static", color: "bg-slate-100 text-slate-700" },
  { value: "piping", label: "Piping", color: "bg-teal-100 text-teal-700" },
  { value: "electrical", label: "Electrical", color: "bg-amber-100 text-amber-700" },
  { value: "instrumentation", label: "Instrumentation", color: "bg-purple-100 text-purple-700" },
  { value: "civil", label: "Civil", color: "bg-orange-100 text-orange-700" },
  { value: "operations", label: "Operations", color: "bg-green-100 text-green-700" },
  { value: "laboratory", label: "Laboratory", color: "bg-cyan-100 text-cyan-700" },
  { value: "multi_discipline", label: "Multi-discipline", color: "bg-pink-100 text-pink-700" },
];

// Simple list of discipline values
export const DISCIPLINE_VALUES = DISCIPLINES.map(d => d.value);

// Simple list of discipline labels (for dropdowns)
export const DISCIPLINE_LABELS = DISCIPLINES.map(d => d.label);

// Get color class for a discipline
export const getDisciplineColor = (discipline) => {
  const found = DISCIPLINES.find(d => 
    d.value.toLowerCase() === (discipline || "").toLowerCase() ||
    d.label.toLowerCase() === (discipline || "").toLowerCase()
  );
  return found?.color || "bg-slate-100 text-slate-700";
};

// Get discipline label from value
export const getDisciplineLabel = (value) => {
  const found = DISCIPLINES.find(d => 
    d.value.toLowerCase() === (value || "").toLowerCase() ||
    d.label.toLowerCase() === (value || "").toLowerCase()
  );
  return found?.label || value;
};

/** Resolve a translated discipline label for UI selects and badges. */
export const translateDiscipline = (t, discipline) => {
  if (!discipline) return "";
  const normalized = normalizeDiscipline(discipline);
  const found = DISCIPLINES.find(
    (d) =>
      d.value === discipline ||
      d.value === normalized ||
      d.label.toLowerCase() === String(discipline).toLowerCase()
  );
  const keys = [
    found?.value,
    found?.label,
    discipline,
    normalized,
    String(discipline).replace(/[\s-]+/g, "_").toLowerCase(),
  ].filter(Boolean);
  const seen = new Set();
  for (const key of keys) {
    if (seen.has(key)) continue;
    seen.add(key);
    const translated = t(`disciplines.${key}`);
    if (typeof translated === "string" && translated !== `disciplines.${key}`) {
      return translated;
    }
  }
  return found?.label || getDisciplineLabel(discipline) || discipline;
};

// Normalize a discipline value (handle legacy/case variations)
export const normalizeDiscipline = (value) => {
  if (!value) return "";
  
  // Direct match by value or label
  const direct = DISCIPLINES.find(d => 
    d.value === value || 
    d.label === value ||
    d.value.toLowerCase() === value.toLowerCase() ||
    d.label.toLowerCase() === value.toLowerCase()
  );
  if (direct) return direct.value;
  
  // Legacy mappings for backward compatibility
  const legacyMap = {
    // Old discipline names mapped to new ones
    "mechanical": "rotating",
    "rotating equipment": "rotating",
    "static equipment": "static",
    "process": "operations",
    "maintenance": "operations",
    "safety": "operations",
    "inspection": "laboratory",
    "lab": "laboratory",
    "reliability": "operations",
    "multi-discipline": "multi_discipline",
    "multi_discipline": "multi_discipline",
    "engineering": "operations",
  };
  
  const mapped = legacyMap[value.toLowerCase()];
  if (mapped) return mapped;
  
  // Return lowercase as fallback
  return value.toLowerCase().replace(/[\s-]/g, "_");
};

File: backend/models/disciplines.py
"""
Standard discipline values used across PM import, tasks, actions, and FMEA.

Aligned with:
- iso14224_models.Discipline
- frontend/src/constants/disciplines.js
- routes/disciplines.py SEED_DISCIPLINES
"""

from enum import Enum
from typing import Dict, List, Optional, Tuple


class StandardDiscipline(str, Enum):
    ROTATING = "rotating"
    STATIC = "static"
    PIPING = "piping"
    ELECTRICAL = "electrical"
    INSTRUMENTATION = "instrumentation"
    CIVIL = "civil"
    OPERATIONS = "operations"
    LABORATORY = "laboratory"


DISCIPLINE_LIST: List[str] = [d.value for d in StandardDiscipline]
DEFAULT_DISCIPLINE = StandardDiscipline.ROTATING.value

# (value, label, aliases) — keep in sync with SEED_DISCIPLINES / frontend constants
_STANDARD_DISCIPLINE_ROWS: Tuple[Tuple[str, str, Tuple[str, ...]], ...] = (
    ("rotating", "Rotating", ("mechanical", "rotating equipment", "mech", "td", "tech", "technisch", "werktuigbouw")),
    ("static", "Static", ("static equipment",)),
    ("piping", "Piping", ()),
    ("electrical", "Electrical", ("e&i", "elec", "elektro")),
    ("instrumentation", "Instrumentation", ("inst", "instr", "i&c")),
    ("civil", "Civil", ()),
    (
        "operations",
        "Operations",
        (
            "process",
            "maintenance",
            "safety",
            "reliability",
            "multi-discipline",
            "multi_discipline",
            "engineering",
            "ops",
            "op",
            "oper",
            "bediening",
            "maint",
            "onderhoud",
            "hvac",
        ),
    ),
    ("laboratory", "Laboratory", ("inspection", "lab")),
)

DISCIPLINE_LABELS: Dict[str, str] = {value: label for value, label, _ in _STANDARD_DISCIPLINE_ROWS}
DISCIPLINE_LABEL_LIST: List[str] = [label for _, label, _ in _STANDARD_DISCIPLINE_ROWS]

_ALIAS_TO_VALUE: Dict[str, str] = {}
for value, label, aliases in _STANDARD_DISCIPLINE_ROWS:
    _ALIAS_TO_VALUE[value] = value
    _ALIAS_TO_VALUE[label.lower()] = value
    _ALIAS_TO_VALUE[label.lower().replace(" ", "_")] = value
    for alias in aliases:
        _ALIAS_TO_VALUE[alias.lower()] = value
        _ALIAS_TO_VALUE[alias.lower().replace(" ", "_").replace("-", "_")] = value


def _lookup_key(value: str) -> str:
    return value.strip().lower().replace(" ", "_").replace("-", "_")


def normalize_discipline(value: Optional[str]) -> str:
    """
    Normalize a discipline string to a canonical lowercase value (e.g. rotating).
    Returns empty string when input is blank or cannot be mapped.
    """
    if not value or not str(value).strip():
        return ""

    raw = str(value).strip()
    for key in (_lookup_key(raw), raw.lower()):
        mapped = _ALIAS_TO_VALUE.get(key)
        if mapped:
            return mapped

    return ""


def normalize_discipline_or_default(value: Optional[str]) -> str:
    """Normalize discipline; fall back to DEFAULT_DISCIPLINE when unknown."""
    return normalize_discipline(value) or DEFAULT_DISCIPLINE
